x/merkledb/trie.go

```diff
@@ -36,52 +70,199 @@ type ReadOnlyTrie interface {
-	getRoot() maybe.Maybe[*node]
-
-	// get an editable copy of the node with the given key path
-	// hasValue indicates which db to look in (value or intermediate)
-	getEditableNode(key Key, hasValue bool) (*node, error)
-
 	// GetRangeProof returns a proof of up to [maxLength] key-value pairs with
 	// keys in range [start, end].
 	// If [start] is Nothing, there's no lower bound on the range.
 	// If [end] is Nothing, there's no upper bound on the range.
 	// Returns ErrEmptyProof if the trie is empty.
 	GetRangeProof(ctx context.Context, start maybe.Maybe[[]byte], end maybe.Maybe[[]byte], maxLength int) (*RangeProof, error)
 
-	database.Iteratee
-}
-
-type ViewChanges struct {
-	BatchOps []database.BatchOp
-	MapOps   map[string]maybe.Maybe[[]byte]
-	// ConsumeBytes when set to true will skip copying of bytes and assume
-	// ownership of the provided bytes.
-	ConsumeBytes bool
-}
-
-type Trie interface {
-	ReadOnlyTrie
-
 	// NewView returns a new view on top of this Trie where the passed changes
 	// have been applied.
 	NewView(
 		ctx context.Context,
 		changes ViewChanges,
-	) (TrieView, error)
+	) (View, error)
 }
 
-type TrieView interface {
+type View interface {
 	Trie
 
 	// CommitToDB writes the changes in this view to the database.
 	// Takes the DB commit lock.
 	CommitToDB(ctx context.Context) error
 }
+
+// Returns the nodes along the path to [key].
+// The first node is the root, and the last node is either the node with the
+// given [key], if it's in the trie, or the node with the largest prefix of
+// the [key] if it isn't in the trie.
+// Always returns at least the root node.
+func visitPathToKey[T Trie](t T, key Key, visitNode func(*node) error) error {
+	maybeRoot := t.getRoot()
+	if maybeRoot.IsNothing() {
+		return nil
+	}
+	root := maybeRoot.Value()
+	if !key.HasPrefix(root.key) {
+		return nil
+	}
+	var (
+		// all node paths start at the root
+		currentNode = root
+		tokenSize   = t.getTokenSize()
+		err         error
+	)
+	if err := visitNode(currentNode); err != nil {
+		return err
+	}
+	// while the entire path hasn't been matched
+	for currentNode.key.length < key.length {
+		// confirm that a child exists and grab its ID before attempting to load it
+		nextChildEntry, hasChild := currentNode.children[key.Token(currentNode.key.length, tokenSize)]
+
+		if !hasChild || !key.iteratedHasPrefix(nextChildEntry.compressedKey, currentNode.key.length+tokenSize, tokenSize) {
+			// there was no child along the path or the child that was there doesn't match the remaining path
+			return nil
+		}
+		// grab the next node along the path
+		currentNode, err = t.getNode(key.Take(currentNode.key.length+tokenSize+nextChildEntry.compressedKey.length), nextChildEntry.hasValue)
+		if err != nil {
+			return err
+		}
+		if err := visitNode(currentNode); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// Returns a proof that [bytesPath] is in or not in trie [t].
+func getProof[T Trie](t T, key []byte) (*Proof, error) {
```

Comment:
These functions are tested indirectly elsewhere, but it would probably be a good idea to test them directly, too